### PR TITLE
refactor: Add a field in Pool to specify the retry times

### LINKED ISSUE #3469: Add a field `n_retries` to `Pool` to specify the number of times to retry when getting a pool-item
Add a field `n_retries` to `Pool` to specify the number of times to retry when getting a pool-item:

- `n_retries` has to be non-zero.
- Add test to ensure this behavior

https://github.com/datafuselabs/databend/blob/9c7b86f841a66f9e8bde7bced234947c7d041179/common/base/src/containers/pool.rs#L78-L85

https://github.com/datafuselabs/databend/blob/9c7b86f841a66f9e8bde7bced234947c7d041179/common/base/src/containers/pool.rs#L131-L157

---

> I prefer `n_retry` as a parameter with default value 3

_Originally posted by @ariesdevil in https://github.com/datafuselabs/databend/pull/3468#discussion_r768890671_

Maybe make it a configureable struct field?

### PR DESCRIPTION
**Pull request by junaire**
Signed-off-by: Jun Zhang <jun@junz.org>

I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Fixes #3469

### CODE REVIEW

**Comment by junaire** [comment]
Few questions:
1. How can we test it?
2. Rust seems not to support default parameters yet, how can we omit the odd `3`? (Sorry new to Rust, I'm a C艹 people)

File: src/common/base/src/containers/pool.rs [hunk 1]
```diff
@@ -75,12 +77,13 @@ where
     Mgr::Item: Clone + Sync + Send + Debug,
     Mgr::Error: Sync + Debug,
 {
-    pub fn new(manager: Mgr, initial_retry_interval: Duration) -> Self {
+    pub fn new(manager: Mgr, initial_retry_interval: Duration, retries: Option<u32>) -> Self {
```
Comment (Xuanwo) [inline src/common/base/src/containers/pool.rs:80]: I prefer to add a new API called `with_retries(self, retries: u32) -> Self` for `Pool`.

- Users who don't care about retry times can use `Pool::new()` AS-IS.
- Users who want to change the default retries can use `Pool::new().with_retries(123)`.

Comment (junaire) [reply src/common/base/src/containers/pool.rs:80]: I love your approach more. I'll update the PR later, not accessible to my work machine now. BTW do you know what's the problem with the CI? Is there something wrong with my code, or just some random glitches?

Comment (Xuanwo) [reply src/common/base/src/containers/pool.rs]: > BTW do you know what's the problem with the CI? Is there something wrong with my code, or just some random glitches?

Sorry for my lateness. I'm going out today.

Based on the error message, I think we need to update the test:

```rust
running 13 tests
test pool::test_pool ... FAILED
test progress::test_progress ... ok
test range_key_test::test_range_wrapper_reflexive_property ... ok
test range_map_test::test_range_set ... ok
test runtime::test_runtime ... ok
test pool_retry::test_pool ... ok
test stoppable::test_stop_handle ... ok
test stoppable::test_stop_handle_drop ... ok
test stoppable::test_stoppable ... ok
test string_func::mask_string_test ... ok
test string_func::replace_nth_char_test ... ok
test string_func::test_progress ... ok
test runtime::test_shutdown_long_run_runtime ... ok

failures:

---- pool::test_pool stdout ----
thread 'pool::test_pool' panicked at 'assertion failed: `(left == right)`
  left: `2`,
 right: `3`: seq=2 valid for make', src/common/base/tests/it/pool.rs:37:5

failures:
    pool::test_pool
```

**Comment by junaire** [comment]
FWIW, I can't reproduce the failure in CI
image
image

**Comment by Xuanwo** [comment]
We added a new test which will call:

```rust
let seq = GlobalSequence::next();
```

I think this makes our pool test flaky.

CC @drmingdrmer for advice.

**Review by Xuanwo — APPROVED** [review]
Nice!

**Comment by Xuanwo** [comment]
Thanks, @junaire, for your contribution!

**Pull request merged**